pkg/networkservice/common/refresh/client.go

```diff
@@ -55,76 +54,67 @@ func (t *refreshClient) Request(ctx context.Context, request *networkservice.Net
 	if err != nil {
 		return nil, err
 	}
-	// Clone the request
-	req := request.Clone()
-	// Set its connection to the returned connection we received
-	req.Connection = rv
-
-	expire, err := t.getExpireDuration(request)
+	expireTime, err := ptypes.Timestamp(request.GetConnection().GetPath().GetPathSegments()[request.GetConnection().GetPath().GetIndex()].GetExpires())
 	if err != nil {
-		return nil, errors.Wrapf(err, "Error creating timer from Request.Connection.Path.PathSegment[%d].ExpireTime", request.GetConnection().GetPath().GetIndex())
+		return nil, err
 	}
+	// TODO - introduce random noise into duration avoid timer lock
+	duration := time.Until(expireTime) / 3
+	trace.Log(ctx).Infof("refreshDuration: %+v", duration)
 	t.executor.AsyncExec(func() {
-		id := req.GetConnection().GetId()
-		// check if it is refresh request
-		if refreshCtx := refreshContext(ctx); refreshCtx != nil {
-			// refresh was canceled
-			if refreshCtx.Err() != nil {
-				return
-			}
-			// we reuse non-canceled refresh context for the next refresh request
-			timer := t.createTimer(ctx, req, expire, opts...)
-			t.connectionTimers[id] = timer
-		} else {
-			// cancel refresh of previous request if any
-			if timer, ok := t.connectionTimers[id]; ok {
-				timer.Stop()
-			}
-			if canceller, ok := t.refreshCancellers[id]; ok {
-				canceller()
-			}
-			// add refresh context to request context
-			refreshCtx, cancelFunc := context.WithCancel(context.Background())
-			newCtx := withRefreshContext(ctx, refreshCtx)
+		// Create refreshRequest
+		refreshRequest := request.Clone()
```

Comment:
I think to avoid potential race condition we need to move this cloning out of async exec section.

Reply:
You are correct, fixed in update :)